browser/component_updater/brave_component_contents_verifier.cc

```diff
@@ -0,0 +1,184 @@
+    0x1e, 0x94, 0x48, 0x0a, 0x54, 0x05, 0x52, 0x01, 0x04, 0xd2, 0xef, 0x57,
+    0x89, 0x47, 0xd3, 0xae, 0xd9, 0xc1, 0xf5, 0xa9, 0xbf, 0x60, 0x96, 0x4b,
+    0xa0, 0x12, 0x9c, 0xf3, 0x00, 0xe6, 0x32, 0x40, 0xc7, 0x6e, 0x29, 0xa8,
+    0x81, 0xfd, 0x2f, 0x1e, 0x92, 0x4a, 0x5e, 0xed, 0xef, 0x13, 0x9f, 0xed,
+    0x88, 0x77, 0x2c, 0x84, 0xdd, 0x00, 0x87, 0x03, 0x49, 0x09, 0xb7, 0x4b,
+    0xc7, 0x02, 0x03, 0x01, 0x00, 0x01};
+
+constexpr char kVerifiedContentsPath[] = "_metadata/verified_contents.json";
+
+std::string GetRootHasheForContent(base::span<const uint8_t> contents,
+                                   size_t block_size) {
+  std::vector<std::string> hashes;
+  // Even when the contents is empty, we want to output at least one hash
+  // block (the hash of the empty string).
+  base::SpanReader reader(contents);
+  do {
+    auto chunk = reader.Read(std::min(block_size, reader.remaining()));
+    std::unique_ptr<crypto::SecureHash> hash(
+        crypto::SecureHash::Create(crypto::SecureHash::SHA256));
+    hash->Update(chunk.value());
+
+    std::string buffer;
+    buffer.resize(crypto::kSHA256Length);
+    hash->Finish(base::as_writable_byte_span(buffer));
+    hashes.push_back(std::move(buffer));
+  } while (reader.remaining() > 0);
+
+  CHECK(block_size % crypto::kSHA256Length == 0);
+  return extensions::ComputeTreeHashRoot(hashes,
+                                         block_size / crypto::kSHA256Length);
+}
+
+// Proxies the file access and checks the verified_contents.json.
+class ComponentContentsAccessorImpl
+    : public ComponentNoChecksContentsAccessorImpl {
+ public:
+  explicit ComponentContentsAccessorImpl(const base::FilePath& component_root)
+      : ComponentNoChecksContentsAccessorImpl(component_root) {
+    verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+        kComponentContentsVerifierPublicKey,
+        component_root.AppendASCII(kVerifiedContentsPath));
+    if (verified_contents_ &&
+        verified_contents_->block_size() % crypto::kSHA256Length != 0) {
+      // Unsupported block size.
+      verified_contents_.reset();
+    }
+  }
+
+  bool IsComponentSignatureValid() const override {
+    return !!verified_contents_.get();
+  }
+
+  void IgnoreInvalidSignature(bool ignore) override {
+    ignore_invalid_signature_ = ignore;
+  }
+
+  bool VerifyContents(const base::FilePath& relative_path,
+                      base::span<const uint8_t> contents) override {
+    if (!IsComponentSignatureValid()) {
+      return ignore_invalid_signature_;
+    }
+
+    if (!verified_contents_->HasTreeHashRoot(relative_path)) {
+      return true;
+    }
+
+    const auto root_hash =
+        GetRootHasheForContent(contents, verified_contents_->block_size());
+    return verified_contents_->TreeHashRootEquals(relative_path, root_hash);
+  }
+
+ private:
+  ~ComponentContentsAccessorImpl() override = default;
+
+  std::unique_ptr<extensions::VerifiedContents> verified_contents_;
```

Comment:
Do we need a `unique_ptr` here? I suspect it could be just:
```
extensions::VerifiedContents verified_contents_;
```

Reply:
Ah, it has to be.

Reply:
It has private c-tor and can only be created by static factory methods